host_core/native/hostcore_wasmcloud_native/src/inv.rs

```diff
@@ -0,0 +1,296 @@
+    /// so an invocation requires a reference to the host (signing) key
+    pub fn new(
+        hostkey: &KeyPair,
+        origin: WasmCloudEntity,
+        target: WasmCloudEntity,
+        op: &str,
+        msg: Vec<u8>,
+    ) -> Invocation {
+        let subject = format!("{}", Uuid::new_v4());
+        let issuer = hostkey.public_key();
+        let target_url = format!("{}/{}", target.url(), op);
+        let claims = Claims::<wascap::prelude::Invocation>::new(
+            issuer.to_string(),
+            subject.to_string(),
+            &target_url,
+            &origin.url(),
+            &invocation_hash(&target_url, &origin.url(), &msg),
+        );
+        Invocation {
+            origin,
+            target,
+            operation: op.to_string(),
+            msg,
+            id: subject,
+            encoded_claims: claims.encode(&hostkey).unwrap(),
+            host_id: issuer,
+        }
+    }
+
+    /// Produces a host-signed invocation that is used to halt anything that can receive invocations. This invocation
+    /// has both an origin and a target of SYSTEM_ACTOR. This has a net effect of making this invocation unroutable
+    /// across a lattice, and therefore can only be produced internally. In other words, a remote host can't fabricate
+    /// a halt invocation and send it to a provider or actor
+    pub fn halt(hostkey: &KeyPair) -> Invocation {
+        let subject = format!("{}", Uuid::new_v4());
+        let issuer = hostkey.public_key();
+        let op = OP_HALT.to_string();
+        let target = WasmCloudEntity::Actor(SYSTEM_ACTOR.to_string());
+        let target_url = format!("{}/{}", target.url(), &op);
+        let claims = Claims::<wascap::prelude::Invocation>::new(
+            issuer.to_string(),
+            subject.to_string(),
+            &target_url,
+            &target.url(),
+            &invocation_hash(&target_url, &target.url(), &[]),
+        );
+        Invocation {
+            origin: target.clone(),
+            target,
+            operation: op,
+            msg: vec![],
+            id: subject,
+            encoded_claims: claims.encode(&hostkey).unwrap(),
+            host_id: issuer,
+        }
+    }
+
+    /// A fully-qualified URL indicating the origin of the invocation
+    pub fn origin_url(&self) -> String {
+        self.origin.url()
+    }
+
+    /// A fully-qualified URL indicating the target of the invocation
+    pub fn target_url(&self) -> String {
+        format!("{}/{}", self.target.url(), self.operation)
+    }
+
+    /// The hash of the invocation's target, origin, and raw bytes
+    pub fn hash(&self) -> String {
+        invocation_hash(&self.target_url(), &self.origin_url(), &self.msg)
+    }
+
+    /// Validates the current invocation to ensure that the invocation claims have
+    /// not been forged, are not expired, etc
+    pub fn validate_antiforgery(&self) -> Result<()> {
```

Comment:
I think this is a good candidate for a Rust test here, to ensure this logic is working as intended as it's one of the more important pieces of auth in wasmcloud

Reply:
:thinking: I was pretty sure there was a test for this in the "old" repo.. I'll see if I can find it.